File: Smallest Number in Infinite Set.java
// https://leetcode.com/problems/smallest-number-in-infinite-set/

class SmallestInfiniteSet {
    
    // Create a TreeSet and a variable to hold the smallest number
    // NOTE: TreeSet is a self-balancing BST - similar to a Red-Black tree
    TreeSet<Integer> set;
    int smallestNum;

    public SmallestInfiniteSet() {
        
        // Initialize class variables
        set = new TreeSet<>();
        smallestNum = 1;
    }
    
    public int popSmallest() {
        
        // OBJECTIVE: Pop smallest element from set or return smallestNum
        
        // If tree isn't empty, pop head element
        if (!set.isEmpty()) {
            return set.pollFirst();
        }
        
        // If tree is empty, return smallest number based on variable
        smallestNum++;
        return smallestNum - 1;
    }
    
    public void addBack(int num) {
        
        // OBJECTIVE: Add num to tree if it doesn't exist
        
        // If num is less than smallestNum and doesn't exist in set, add it to set
        if (num < smallestNum && !set.contains(num)) {
            set.add(num);
        }
    }
}

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet obj = new SmallestInfiniteSet();
 * int param_1 = obj.popSmallest();
 * obj.addBack(num);
 */

File: Smallest Number in Infinite Set.py
# https://leetcode.com/problems/smallest-number-in-infinite-set/

import heapq

class SmallestInfiniteSet:
    
    def __init__(self):
        
        # Create a heap and a set
        self.heap = list()
        self.set = set()
        
        # Create a variable to hold the smallest number
        self.smallestNum = 1

    def popSmallest(self) -> int:
        
        # OBJECTIVE: Return smallest element from heap and set
        
        # Check if heap isn't empty
        if self.heap:
            
            # Pop element from heap and set
            popped = heapq.heappop(self.heap)
            self.set.remove(popped)
            
            # Return popped element
            return popped
        
        # If heap is empty, return smallest number based on variable
        # NOTE: smallestNum needs to be updated before returned
        self.smallestNum += 1
        return self.smallestNum - 1

    def addBack(self, num: int) -> None:
        
        # If num is smaller than smallest and it doesn't exist in set, add it to set and heap
        if num < self.smallestNum and num not in self.set:
            self.set.add(num)
            heapq.heappush(self.heap, num)

# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)
